src/api.mjs

```diff
@@ -162,14 +154,54 @@ async function route(message) {
       return { ...message, error: data.errors[0].message };
     }
 
+    return { ...message, results: data };
+  } else if (type === "ipfs") {
+    let { url, gateway, timeout: timeoutFromMsg } = message.options;
+
+    const nativeIPFSPattern = /^(ipfs:\/\/)([^/?#]+)(.*)/;
+    const match = url.match(nativeIPFSPattern);
+
+    if (match === null) return { ...message, error: "Invalid IPFS URL" };
+    const [_, protocol, hash, path] = match;
+
+    if (!protocol) return { ...message, error: "Invalid protcol" };
+    if (!hash) return { ...message, error: "Could not find CID" };
+
+    try {
+      CID.parse(hash);
+    } catch (error) {
+      return { ...message, error: "Invalid CID" };
+    }
+
+    url = `${gateway}${hash}${path}`; // gateway will contain a trailing slash
+
+    const { origin } = new URL(url);
+    const { rateLimiter, timeout: timeoutFromConfig } =
+      endpointStore.get(origin) ?? {};
+
+    if (rateLimiter) {
+      await rateLimiter.removeTokens(1);
+    }
+
+    let signal;
+    if (timeoutFromMsg || timeoutFromConfig) {
+      signal = AbortSignal.timeout(timeoutFromMsg ?? timeoutFromConfig);
+    }
+
+    let data;
+    try {
+      data = await request(url, "GET", null, null, signal);
```

Comment:
a more self-documenting way of writing this is by defining `const body = null` etc and then passing them into the request function